OpInfo: Allow sample_inputs_func to be any iterable

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #69257
* __->__ #69256

Closes #52486

Differential Revision: [D32942008](https://our.internmc.facebook.com/intern/diff/D32942008)

peterbell10 committed:
OpInfo: Allow sample_inputs_func to be any iterable

Closes #52486

[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Allow sample_inputs_func to be any iterable"


Closes #52486

[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Allow sample_inputs_func to be any iterable"


Closes #52486

[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Allow sample_inputs_func to be any iterable"


Closes #52486

[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Allow sample_inputs_func to be any iterable"


Closes #52486

[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -4973,9 +4965,10 @@ def gen_inputs():
-                    B.requires_grad_(grad_B)
-                    yield SampleInput(A, args=(B,), kwargs=kwargs)
+                    yield SampleInput(
+                        A.clone().requires_grad_(grad_A),
```
Why is the clone needed here?

peterbell10 replied:
@mruberry GitHub cuts this off in the diff, but it's inside a loop
```
for grad_A, grad_B in product((True, False), repeat=2):
  # Either A or B needs to have a gradient
  if not grad_A and not grad_B:
    continue
  A.requires_grad_(grad_A)
  B.requires_grad_(grad_B)
  yield SampleInput(A, args=(B,), kwargs=kwargs)
```

So previously the same tensor was returned in multiple `SampleInputs` simultaneously which causes this assert to fail
https://github.com/pytorch/pytorch/blob/fc8404b5bc7e9721aa93021cdc27de818df64d8d/test/test_ops.py#L251

nikitaved replied:
This clone is required to make inputs independent yet with the same content.

mruberry commented on torch/testing/_internal/common_utils.py:
```diff
@@ -3103,3 +3103,10 @@ def measure() -> float:
     return mean(vals[2 : num - 2])
+
+
+def first_sample(self, samples):
```
Add a comment documenting this function

peterbell10 replied:
Done

mruberry reviewed (approved):
Thanks @peterbell10!

See two minor questions/requests inline

mruberry commented:
@pytorchbot ciflow rerun -l ciflow/all

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32942008).

Lezcano commented:
@nikitaved the day has come.

nikitaved commented:
There is an outstanding PR: https://github.com/pytorch/pytorch/pull/53044. When compared to the one by @peterbell10 , it does not materialize the full sequence at once. Maybe it is better now since we have several wrappers which change the global random state without restoring it.

peterbell10 committed:
Update on "OpInfo: Allow sample_inputs_func to be any iterable"


Closes #52486

Differential Revision: [D32942008](https://our.internmc.facebook.com/intern/diff/D32942008)

[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Allow sample_inputs_func to be any iterable"


Closes #52486

Differential Revision: [D32942008](https://our.internmc.facebook.com/intern/diff/D32942008)

[ghstack-poisoned]

peterbell10 commented:
>When compared to the one by @peterbell10 , it does not materialize the full sequence at once.

I only call `list(samples)` in a few tests that iterate over the samples multiple times. There are still many test cases that iterate over the samples one at a time.

nikitaved commented:
> I only call `list(samples)` in a few tests that iterate over the samples multiple times. There are still many test cases that iterate over the samples one at a time.

I see, so there are tests which follow the processing like this `sample1 -> op(sample1) -> sample2 -> op(sample2) -> ...`? If so, we need to enable the fixture that is commented out in https://github.com/pytorch/pytorch/pull/69107#issuecomment-988095274, not not alter the global random state.

peterbell10 committed:
Update on "OpInfo: Allow sample_inputs_func to be any iterable"


Closes #52486

Differential Revision: [D32942008](https://our.internmc.facebook.com/intern/diff/D32942008)

[ghstack-poisoned]

peterbell10 committed:
Fix sparse csr tests on "OpInfo: Allow sample_inputs_func to be any iterable"


Closes #52486

Differential Revision: [D32942008](https://our.internmc.facebook.com/intern/diff/D32942008)

[ghstack-poisoned]

peterbell10 commented:
I managed to get `freeze_rng_state` working with `test_composite_compliance`.

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32942008).

peterbell10 committed:
Fix merge conflict on "OpInfo: Allow sample_inputs_func to be any iterable"


Closes #52486

Differential Revision: [D32942008](https://our.internmc.facebook.com/intern/diff/D32942008)

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32942008).

peterbell10 commented:
Master is failing because #69399 added a test that indexes into `sample_inputs`. #69697 will fix that test.

zou3519 commented on torch/testing/_internal/common_utils.py:
```diff
@@ -1123,13 +1124,21 @@ def set_rng_seed(seed):
-    if torch.cuda.is_available():
-        torch.cuda.set_rng_state(cuda_rng_state)
-    torch.set_rng_state(rng_state)
+    # no_dispatch needed for test_composite_compliance
```
Hmm, so the implication here is that torch.get_rng_state and torch.set_rng_state are not composite compliant, right?

peterbell10 replied:
No, the problem is that `Generator.set_rng_state()` doesn't support `__torch_dispatch__` which is used to implement the test.
